File: C++/1076-forwards_on_weibo.cc
#include <iostream>
#include <vector>
#include <queue>


using Graph = std::vector<std::vector<int>>;


int maximum_potential_forwards(const Graph &graph, int max_level, int query) {
    int count = 0;
    std::vector<bool> visited(graph.size());
    std::vector<int> dist(graph.size());
    std::queue<int> queue;
    visited[query] = true;
    queue.push(query);
    while (!queue.empty()) {
        int u = queue.front();
        queue.pop();
        for (auto &v : graph[u]) {
            if (!visited[v]) {
                dist[v] = dist[u] + 1;
                if (dist[v] < max_level) {
                    ++count;
                    queue.push(v);
                } else if (dist[v] == max_level) {
                    ++count;
                }
                visited[v] = true;
            }
        }
    }
    return count;
}


int main() {

    std::ios::sync_with_stdio(false);

    int n_node, max_level;
    std::cin >> n_node >> max_level;

    Graph graph(n_node + 1);
    for (int u=1; u<=n_node; ++u) {
        int n_following;
        std::cin >> n_following;
        while (n_following--) {
            int v;
            std::cin >> v;
            graph[v].push_back(u);
        }
    }

    int n_query;
    std::cin >> n_query;
    while (n_query--) {
        int query;
        std::cin >> query;
        std::cout << maximum_potential_forwards(graph, max_level, query) << '\n';
    }

    return 0;
}

File: Python/1076-forwards_on_weibo.py
from collections import deque


def maximum_potential_forwards(query, graph, max_level):
    count = 0
    visited = [False] * len(graph)
    dist = [0] * len(graph)
    visited[query] = True
    queue = deque([query])
    while queue:
        u = queue.popleft()
        for v in graph[u]:
            if visited[v] is False:
                dist[v] = dist[u] + 1
                if dist[v] < max_level:
                    count += 1
                    queue.append(v)
                elif dist[v] == max_level:
                    count += 1
                visited[v] = True
    return count


if __name__ == '__main__':
    n_node, max_level = [int(i) for i in input().split()]
    graph = [[] for _ in range(n_node + 1)]
    for u in range(1, n_node + 1):
        user_list = [int(i) for i in input().split()][1:]
        for v in user_list:
            graph[v].append(u)
    query_list = [int(i) for i in input().split()][1:]
    for query in query_list:
        print(maximum_potential_forwards(query, graph, max_level))
